### DISCUSSION TITLE
Should zone filter `threshold` actually be `min_score`?

### DISCUSSION THREAD

**Discussion by qvr** [blakeblackshear/frigate#8938] [General]
I was going to open a PR directly, but there's a couple of different ways I can see this issue could be solved, so opening a discussion first. :)

Basically setting `threshold` in zone filters does not do what `threshold` in general does, ie. the score limit for triggering an event. In zone filter [the setting](https://github.com/blakeblackshear/frigate/blob/dev/frigate/object_processing.py#L435) is [only used to decide whether to keep zone history](https://github.com/blakeblackshear/frigate/blob/dev/frigate/object_processing.py#L194) for that object.

I would like to keep zone history for an object once it goes over the `min_score` setting so Frigate has started to track it, so once it passes the `threshold` limit and triggers an event it would already have a zone history of from where the object entered the frame/frigate started tracking it, instead of where it was decided it was no longer a false positive. This would seem to make sense, as if we have a setting defining the score to track an object, should tracking involve zone history as well?

Because the `threshold` in zone filters actually does not matter at all when it comes to [false positive decision](https://github.com/blakeblackshear/frigate/blob/dev/frigate/object_processing.py#L161), I can already just set this threshold to a lower value and the history would be stored. This is not however very intuitive, and other users will probably not easily realise this, but I guess a documentation fix could be helpful.

The `zone_filtered` function could also be altered to check only `min_score`, or `min(min_score,threshold)` to keep it somewhat backwards compatible.

Thoughts?

**Comment by NickM-27** [comment]
There is a reason that the threshold and min_score filters are not documented at the zone level except in the full config reference - in general they should not be used.

You are correct that setting threshold at the zone filter level simply filters if that object can be in the zone and by default it is not set at all and that is what I would recommend leaving it as.

**Reply by qvr** [reply]
> by default it is not set at all

Hmm, but as far as I can tell, by default it uses the[ inherited object level](https://github.com/blakeblackshear/frigate/blob/dev/frigate/config.py#L448) `threshold` setting, which [defaults to a higher value](https://github.com/blakeblackshear/frigate/blob/dev/frigate/config.py#L385) than min_score?

Which then causes this issue that by default frigate will not _track_ zone history for an object is has decided to track. :)

**Reply by NickM-27** [reply]
you are misunderstanding this code. The `default_factory` is `dict` which means that by default the `filters` is an empty dict that is ONLY set when the user explicitly sets a filter on zone. So by default the value is not set at all.

We can double check this by looking at `http://frigate_ip:5000/api/config` where for example on my config it is very clear how this is behaving

```json
      "zones": {
        "back_deck": {
          "coordinates": "0,720,527,720,298,195,0,275",
          "filters": {},
          "inertia": 3,
          "objects": []
        },
        "back_yard": {
          "coordinates": "1115,278,1251,461,966,720,531,720,396,451,305,193,0,272,0,139,389,103,960,143",
          "filters": {},
          "inertia": 3,
          "objects": []
        },
        "side_yard": {
          "coordinates": "1043,220,1243,464,1280,423,1280,296,1248,257",
          "filters": {},
          "inertia": 16,
          "objects": []
        }
      }
    },
```

**Reply by qvr** [reply]
Well that's interesting, and a head-scratcher, as I'm seeing events which seem to indicate exactly this scenario (`new` events with `before` payload having score > min_score but no zones, and `after` payload with score > threshold and with zone information). I guess I'll keep digging...

Thanks for your help!

**Reply by NickM-27** [reply]
what is the problem with that?

**Reply by qvr** [reply]
I have my zones setup so that if an object is tracked, it *should* be on some zone so those kind of event payloads with no zone information should not be possible.

The problem this then causes is that I have automation based on from where an object (person) initially entered the frame, and mostly it works but then I have these situations where it seems a person has magically jumped a zone.

Of course it can be a bunch of things causing this, but reading the code this just looked like a smoking gun.

**Reply by NickM-27** [reply]
I see, it is worth noting that the /events topic does not care about required zones, it is the raw information about an object so the `before` could occur before an object has entered a zone since that is the state of the object when it was first detected

**Reply by qvr** [reply]
As a test I now defined a zone that covers the entire frame, so basically `0,360,640,360,640,0,0,0` on a 640x360 feed, and inertia is 1 (with my fix applied). I still see `before` events without any zone information.

I'm having a hard time to see what code path could end up like this, any pointers? Or is this to be expected somehow?

**Reply by blakeblackshear** [reply]
The very first message for an object? If so, that would be expected. Only true positive events are considered for zones so that false positives don't trigger all the zone sensors. You should be able to see the change to true positive in the before->after.

**Reply by qvr** [reply]
What threw me was that the before section already contained a score even above threshold, but with no zones, and false_positive:
```
    before:
      id: 1702451354.331449-f836ah
      camera: entrance
      frame_time: 1702451354.331449
      snapshot: null
      label: person
      sub_label: null
      top_score: 0
      false_positive: true
      start_time: 1702451354.331449
      end_time: null
      score: 0.8046875
      current_zones: []
      entered_zones: []
...
    after:
      id: 1702451354.331449-f836ah
      camera: entrance
      frame_time: 1702451354.525434
      label: person
      sub_label: null
      top_score: 0.82421875
      false_positive: false
      score: 0.84375
      current_zones:
        - front_door
        - in_entrance
```

But it looks like [when the TrackedObject](https://github.com/blakeblackshear/frigate/blob/dev/frigate/object_processing.py#L607) is created, it is given the basic object label and scoring which then becomes the initial "before".

It also looks like the false positive decision and zone checks etc aare done only from the second frame onwards, is this correct?

**Reply by qvr** [reply]
Basically, shouldn't [this list of updated_ids](https://github.com/blakeblackshear/frigate/blob/dev/frigate/object_processing.py#L619) include new_ids also?

**Reply by NickM-27** [reply]
for one I just want to point out that `dev` is 0.13 which has had many changes in this regard, I am not sure if you indicated what version you are running.

**Reply by qvr** [reply]
Oh, sorry, I'm on 0.13 beta 6 with [my inertia fix](https://github.com/blakeblackshear/frigate/pull/8890) applied (which is why I started looking into this deeper).

**Reply by qvr** [reply]
> Basically, shouldn't this list of updated_ids include new_ids also?

Opened a PR for this, this fixes the issue I've been seeing:
https://github.com/blakeblackshear/frigate/pull/8951
